pkg/ebpf/events/derive/derive.go

```diff
@@ -0,0 +1,61 @@
+package derive
+
+import (
+	"fmt"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// DeriveFunction is a function prototype for a function that receives an event as
+// argument and may produce a new event if relevant.
+// It returns list of derived events, which might be empty depending on successful derivation
+// and an error if one occurred.
+type DeriveFunction func(trace.Event) ([]trace.Event, error)
+
+// EventSkeleton is a struct for the necessary information from an event definition to create the derived event
+type EventSkeleton struct {
+	Name   string
+	ID     int
+	Params []trace.ArgMeta
+}
+
+// DerivationTable defines a table between events and events they can be derived into corresponding to some deriveFn
+type DerivationTable map[int32]map[int32]DeriveFunction
+
+// Derive takes a trace.Event and checks if it can derive additional events from it
+// as defined by tracee's eventDerivations map.
+// The map is initialized in the above function
+func Derive(event trace.Event, derivationTable DerivationTable) ([]trace.Event, []error) {
+	derivatives := []trace.Event{}
+	errors := []error{}
+	deriveFns := derivationTable[int32(event.EventID)]
+	for id, deriveFn := range deriveFns {
+
+		derivative, err := deriveFn(event)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("failed to derive event %d: %v", id, err))
+		} else if len(derivative) > 0 {
+			derivatives = append(derivatives, derivative...)
+		}
+	}
+
+	return derivatives, errors
+}
+
+// NewEvent create a new derived event from given event values, adjusted by the derived event skeleton meta-data.
+// This method enables using the context of the base event, but with the new arguments and meta-data of the derived one.
+func NewEvent(baseEvent *trace.Event, skeleton EventSkeleton, argsValues []interface{}) (trace.Event, error) {
```

Comment:
Nice! But why pass the event by reference and not value?

Reply:
Just to avoid futile copying of the event, because it is a large struct.